In [174]:
import pandas as pd
import os
from sklearn.metrics import classification_report, accuracy_score

In [293]:
# ["AWPNLI", "RedditNLI", "StressTest", "RTE_Quant", "NewsNLI"]
dataset = "NewsNLI"
root_path = os.path.dirname(os.getcwd())
results_path = os.path.join(root_path, "data", "equate_labelled", f"{dataset}_phase2.csv")
os.makedirs(os.path.join(root_path, "data", "equate_labelled", "processed"), exist_ok=True)

In [294]:
res = pd.read_csv(results_path)
res.head()

sample_index generated_label  error_message golden_label  \
0           575      entailment            NaN      neutral   
1           914   contradiction            NaN      neutral   
2           450      entailment            NaN   entailment   
3           860         neutral            NaN      neutral   
4           831         neutral            NaN      neutral   

                                             premise  \
0  Peter Siebold, 43, was identified as the pilot...   
1  Holder said the Justice Department is working ...   
2  The nine were brought into Somali waters and t...   
3  There are some additional 863 cancer cases amo...   
4  The survey said 57% supported it going forward...   

                                          hypothesis  
0  Pilot who survived identified as Peter Siebold...  
1  Ten Mexican nationals have been charged and se...  
2  Nine people were brought into Somali waters, h...  
3  There are at least 2,509 certified cancer case...  
4  New CNN/ORC survey shows 57% of Americans supp...

In [274]:
res.shape

(247, 6)

In [275]:
res[res.duplicated(subset=["sample_index"])].shape[0]

0

In [246]:
res = res.drop_duplicates(subset=['sample_index'], keep='first', ignore_index=True)
res.to_csv(results_path, index=False)
res.shape

(958, 6)

In [209]:
# import re
#
# def clean_text(text: str):
#     return re.sub(r'\s+', ' ', text.lower().replace("\n", "")).strip()

In [47]:
# for col in ["premise", "hypothesis"]:
#     res[f"clean_{col}"] = res[col].apply(lambda text: clean_text(text))
#
# # mark correctly classified samples and put them first in the dataset, so they don't get drop if they have a duplicate
# res["keep"] = res.apply(lambda row: 0 if row["golden_label"] == row["generated_label"] else 1, axis=1)
# res = res.sort_values(by=["keep"], ascending=True)
# res = res[~res.duplicated(subset=['clean_premise', 'clean_hypothesis'])]
# res.shape

In [48]:
# res.drop(columns=["keep"], axis=0, inplace=True)

In [49]:
# res.to_csv(os.path.join(root_path, "data", "equate_labelled", f"cleaned_{dataset}_gpt4.csv"), index=False)

Do we have scripts that threw an error?

In [247]:
res[res['error_message'].notna()].shape[0]

0

In [248]:
list(res[res['error_message'].notna()]["sample_index"].values)

[]

Filter out erroneous scripts, if any

In [295]:
valid_res = res[res["error_message"].isna()]

In [289]:
valid_res["golden_label"].value_counts(normalize=True)

golden_label
neutral       0.580247
entailment    0.419753
Name: proportion, dtype: float64

In [290]:
valid_res["generated_label"].value_counts()

generated_label
entailment       87
neutral          49
contradiction    26
Name: count, dtype: int64

In [270]:
# valid_res = valid_res[valid_res["generated_label"]!="neutral"]

In [262]:
valid_res[valid_res["generated_label"]=="contradiction"]

sample_index generated_label  error_message golden_label  \
1             914   contradiction            NaN      neutral   
17            736   contradiction            NaN      neutral   
43            935   contradiction            NaN      neutral   
44            727   contradiction            NaN      neutral   
55            663   contradiction            NaN      neutral   
78            749   contradiction            NaN   entailment   
80             64   contradiction            NaN   entailment   
139           838   contradiction            NaN      neutral   
157           818   contradiction            NaN      neutral   
196           605   contradiction            NaN      neutral   
230           787   contradiction            NaN      neutral   
247           586   contradiction            NaN      neutral   
248           681   contradiction            NaN      neutral   
271           930   contradiction            NaN      neutral   
278           844   contradiction            NaN      neutral   
294           835   contradiction            NaN      neutral   
323           726   contradiction            NaN      neutral   
335           896   contradiction            NaN      neutral   
340           737   contradiction            NaN      neutral   
351           643   contradiction            NaN      neutral   
372           505   contradiction            NaN   entailment   
390           540   contradiction            NaN      neutral   
397           762   contradiction            NaN      neutral   
409           931   contradiction            NaN   entailment   
438           293   contradiction            NaN   entailment   
496           173   contradiction            NaN   entailment   
518            24   contradiction            NaN   entailment   
523           498   contradiction            NaN      neutral   
528           659   contradiction            NaN      neutral   
533            96   contradiction            NaN   entailment   
565           669   contradiction            NaN      neutral   
604           918   contradiction            NaN      neutral   
646           829   contradiction            NaN      neutral   
670           257   contradiction            NaN   entailment   
687           879   contradiction            NaN      neutral   
698           111   contradiction            NaN   entailment   
725           859   contradiction            NaN      neutral   
746           765   contradiction            NaN      neutral   
748           734   contradiction            NaN      neutral   
784           953   contradiction            NaN      neutral   
800           856   contradiction            NaN      neutral   
833           553   contradiction            NaN      neutral   
849           857   contradiction            NaN      neutral   
854           705   contradiction            NaN      neutral   
859           576   contradiction            NaN      neutral   
867           890   contradiction            NaN      neutral   
913           750   contradiction            NaN      neutral   
914           853   contradiction            NaN      neutral   
922           523   contradiction            NaN      neutral   
929           189   contradiction            NaN   entailment   
932           562   contradiction            NaN      neutral   
935           600   contradiction            NaN      neutral   
936           740   contradiction            NaN      neutral   
941           812   contradiction            NaN      neutral   
943           725   contradiction            NaN      neutral   

                                               premise  \
1    Holder said the Justice Department is working ...   
17   But suddenly the two reporters were pinned aga...   
43   Despite design improvements, warnings and inst...   
44   Furyk took full advantage to record six birdie...   
55   Youth unemployment is at 19.9 percent in the e...   
78   '' Even if 

In [263]:
valid_res[valid_res["generated_label"]=="contradiction"].to_excel("NewsNLI_predicted_contradiction.xlsx")

In [216]:
valid_res["golden_label"].value_counts(normalize=True)

golden_label
entailment       0.333909
neutral          0.333909
contradiction    0.332181
Name: proportion, dtype: float64

In [ ]:
# # AWPNLI baselines
# mb_baseline_accuracy = 0.50
# rsb_baseline_accuracy = 0.50
# rsb_baseline_f1_c = 0.5
# rsb_baseline_f1_e = 0.5
#
# # RedditNLI baselines
# mb_baseline_accuracy = 57.89
# rsb_baseline_accuracy = 45.94
# rsb_baseline_f1_e = 57.89
# rsb_baseline_f1_n = 34.41
# rsb_baseline_f1_n = 7.69
#
# # StressTest baselines
# mb_baseline_accuracy = 0.33
# rsb_baseline_accuracy = 0.33
# rsb_baseline_f1_c = 0.33
# rsb_baseline_f1_e = 0.33
# rsb_baseline_f1_n = 0.33
#
# # NewsNLI baselines
# mb_baseline_accuracy = 0
# rsb_baseline_accuracy = 0
# rsb_baseline_f1_c = 0
# rsb_baseline_f1_e = 0
#
# # RTE_Quant baselines
# mb_baseline_accuracy = 0
# rsb_baseline_accuracy = 0
# rsb_baseline_f1_c = 0
# rsb_baseline_f1_e = 0

In [217]:
valid_res[valid_res['golden_label'] == valid_res['generated_label']].shape[0]

4658

In [218]:
accuracy_score(y_true=valid_res['golden_label'], y_pred=valid_res['generated_label'])

0.6706983441324694

In [219]:
print(classification_report(y_true=valid_res['golden_label'], y_pred=valid_res['generated_label']))

               precision    recall  f1-score   support

contradiction       0.81      0.70      0.75      2307
   entailment       0.57      0.62      0.60      2319
      neutral       0.66      0.69      0.67      2319

     accuracy                           0.67      6945
    macro avg       0.68      0.67      0.67      6945
 weighted avg       0.68      0.67      0.67      6945



In [285]:
valid_res = valid_res[["sample_index", "premise", "hypothesis", "golden_label", "generated_label"]].rename(columns={"generated_label": "reference_label"})
valid_res.to_csv(os.path.join(root_path, "data", "equate_labelled", "processed", f"{dataset}.csv"), index=False)

## Only for RTE_Quant and NEWSNLI

In [229]:
# valid_res[valid_res["generated_label"]=="contradiction"][["sample_index", "premise", "hypothesis"]].to_excel("NewsNLI_predicted_contradiction.xlsx")

# target label is neutral, model labels as contradiction and manual inspection indicates contradiction as well  (only applicable to NewsNLI and RTE_Quant, which treat "contradiction" case as neutral)
news_nli_corrected_labels_indices = [914, 736, 663, 749, 838, 605]
mask_news = valid_res['sample_index'].isin(news_nli_corrected_labels_indices)

rte_quant_corrected_labels_indices = [118, 128, 98, 7, 21, 54, 147, 20, 30, 155, 52, 95, 120, 72, 150, 57, 111, 26, 115, 58, 4, 18]
mask_rte = valid_res['sample_index'].isin(rte_quant_corrected_labels_indices)

valid_res['corrected_golden_label'] = valid_res["golden_label"]
valid_res.loc[mask_news, 'corrected_golden_label'] = 'contradiction'
valid_res["corrected_golden_label"].value_counts()

corrected_golden_label
neutral          72
entailment       68
contradiction    22
Name: count, dtype: int64

#### First consider the case when we correct the reference label through manual verification. Now the dataset is framed as a 3-way decision.

In [230]:
valid_res["corrected_golden_label"].value_counts(normalize=True)

corrected_golden_label
neutral          0.444444
entailment       0.419753
contradiction    0.135802
Name: proportion, dtype: float64

In [234]:
valid_res[valid_res['corrected_golden_label'] == valid_res['generated_label']].shape[0]

134

In [231]:
accuracy_score(y_true=valid_res['corrected_golden_label'], y_pred=valid_res['generated_label'])

0.8271604938271605

In [235]:
print(classification_report(y_true=valid_res['corrected_golden_label'], y_pred=valid_res['generated_label']))

               precision    recall  f1-score   support

contradiction       0.85      1.00      0.92        22
   entailment       0.75      0.96      0.84        68
      neutral       0.96      0.65      0.78        72

     accuracy                           0.83       162
    macro avg       0.85      0.87      0.84       162
 weighted avg       0.85      0.83      0.82       162



#### Now consider the case when we consider "contradiction" as neutral label, to remain consistent with the results from other papers

In [296]:
valid_res["modified_label"] = valid_res["generated_label"].apply(lambda label: label if label == "entailment" else "neutral")

In [253]:
valid_res['golden_label'].value_counts(normalize=True)

golden_label
entailment    0.504175
neutral       0.495825
Name: proportion, dtype: float64

In [254]:
valid_res[valid_res['golden_label'] == valid_res['modified_label']].shape[0]

696

In [255]:
accuracy_score(y_true=valid_res['golden_label'], y_pred=valid_res['modified_label'])

0.7265135699373695

In [256]:
print(classification_report(y_true=valid_res['golden_label'], y_pred=valid_res['modified_label']))

              precision    recall  f1-score   support

  entailment       0.71      0.77      0.74       483
     neutral       0.74      0.68      0.71       475

    accuracy                           0.73       958
   macro avg       0.73      0.73      0.73       958
weighted avg       0.73      0.73      0.73       958



In [297]:
save_data = valid_res[["sample_index", "premise", "hypothesis", "golden_label", "modified_label"]].rename(columns={"modified_label": "reference_label"})

save_data.to_csv(os.path.join(root_path, "data", "equate_labelled", "processed", f"{dataset}.csv"), index=False)

In [111]:
valid_res.sort_values(by="sample_index").head(10)

sample_index       label  error_message golden_label
134             0     neutral            NaN      neutral
148             1  entailment            NaN      neutral
8               2  entailment            NaN      neutral
11              3  entailment            NaN   entailment
140             4     neutral            NaN      neutral
145             5  entailment            NaN      neutral
5               6  entailment            NaN   entailment
17              7     neutral            NaN      neutral
39              8  entailment            NaN   entailment
54              9  entailment            NaN   entailment

In [112]:
misclassified_samples_indices = sorted(valid_res[valid_res["generated_label"] != valid_res["golden_label"]]["sample_index"].unique())
misclassified_samples_indices

[1,
 2,
 5,
 10,
 20,
 26,
 31,
 37,
 39,
 42,
 47,
 54,
 55,
 65,
 66,
 77,
 81,
 82,
 87,
 89,
 92,
 93,
 94,
 103,
 106,
 109,
 110,
 139,
 142,
 151,
 162]

AWPNLI:
approx issues: 58, 60, 500, 602, 614, 648, 686
ambiguity: 62, 82, 304, 305, 482
wrong label in equate: 107, 106, 109, 138, 550, 674
RedditNLI:
not sure how to correct: 24,
wrong label: 36
RTE_Quant:
wrong label: 26, 31
not sure how to correct: 142

In [113]:
len(misclassified_samples_indices)

31

In [155]:
scripts_path = os.path.join(root_path, "data", "generated", dataset, "script_with_cot_vars_first")
sample_indices = []
for script_file in os.listdir(scripts_path):
    if script_file.endswith(".py"):
        with open(os.path.join(scripts_path, script_file), 'r') as f:
            idx = int(script_file.split(".")[0].split("_")[-1])
            lines = f.readlines()
            inputs = "\n".join(lines[:3]).strip()
            # script = "\n".join(lines)
            if "more than" in inputs or "less than" in inputs:
                sample_indices.append(idx)
sorted(sample_indices)

[0,
 1,
 2,
 3,
 4,
 6,
 7,
 8,
 9,
 10,
 12,
 13,
 14,
 15,
 16,
 18,
 19,
 20,
 21,
 22,
 24,
 26,
 28,
 29,
 30,
 32,
 34,
 35,
 36,
 38,
 39,
 40,
 41,
 42,
 44,
 46,
 47,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 60,
 62,
 64,
 66,
 67,
 68,
 70,
 71,
 72,
 73,
 74,
 76,
 78,
 79,
 80,
 82,
 84,
 86,
 88,
 90,
 92,
 93,
 94,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 106,
 108,
 110,
 111,
 112,
 113,
 116,
 118,
 119,
 120,
 121,
 123,
 124,
 126,
 128,
 130,
 132,
 133,
 134,
 135,
 136,
 138,
 139,
 140,
 142,
 143,
 144,
 146,
 148,
 149,
 150,
 152,
 154,
 155,
 156,
 157,
 158,
 160,
 161,
 162,
 164,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 176,
 177,
 178,
 180,
 182,
 183,
 184,
 186,
 187,
 188,
 190,
 192,
 194,
 195,
 196,
 197,
 198,
 200,
 201,
 202,
 203,
 204,
 205,
 206,
 208,
 209,
 210,
 212,
 214,
 216,
 217,
 218,
 219,
 220,
 221,
 222,
 224,
 226,
 227,
 228,
 230,
 231,
 232,
 234,
 236,
 237,
 238,
 240,
 242,
 243,
 244,
 245,
 2

In [158]:
misclassified_with_quantifier = [idx for idx in misclassified_samples_indices if idx in sample_indices]
len(misclassified_with_quantifier)

130

In [159]:
misclassified_with_quantifier

[1,
 7,
 13,
 15,
 21,
 29,
 35,
 39,
 47,
 55,
 57,
 67,
 68,
 73,
 93,
 96,
 97,
 99,
 101,
 103,
 111,
 113,
 119,
 123,
 124,
 126,
 133,
 135,
 143,
 149,
 155,
 161,
 171,
 173,
 177,
 183,
 190,
 195,
 197,
 203,
 205,
 209,
 217,
 221,
 227,
 237,
 243,
 249,
 259,
 267,
 269,
 271,
 273,
 281,
 283,
 287,
 295,
 311,
 321,
 325,
 327,
 331,
 335,
 337,
 340,
 347,
 368,
 371,
 373,
 383,
 385,
 389,
 392,
 393,
 405,
 415,
 427,
 431,
 439,
 451,
 452,
 473,
 477,
 489,
 491,
 495,
 499,
 501,
 505,
 506,
 509,
 511,
 525,
 527,
 533,
 549,
 577,
 579,
 585,
 589,
 595,
 598,
 601,
 602,
 605,
 611,
 613,
 623,
 629,
 641,
 647,
 649,
 653,
 661,
 663,
 669,
 680,
 683,
 685,
 691,
 693,
 699,
 1168,
 1483,
 2386,
 2727,
 3937,
 4319,
 4873,
 4945]